src/USER-HSMA/hsma2d.cpp

```diff
@@ -0,0 +1,2309 @@
+
+ See the README file in the top-level LAMMPS directory.
+------------------------------------------------------------------------- */
+
+/* ----------------------------------------------------------------------
+   Contributing author: Jiuyang Liang (liangjiuyang@sjtu.edu.cn)
+------------------------------------------------------------------------- */
+
+#include "hsma2d.h"
+
+#include "atom.h"
+#include "comm.h"
+#include "complex.h"
+#include "domain.h"
+#include "error.h"
+#include "force.h"
+#include "math.h"
+#include "math_special.h"
+#include "memory.h"
+#include "update.h"
+
+#include <cstdio>
+#include <cstring>
+#include <cstdlib>
+#include<iomanip>
+#include <immintrin.h>
+#include<iostream>
+
+#if defined(_OPENMP)
+	#include<omp.h>
+#endif
+
+extern "C" {void lfmm3d_t_c_g_(double* eps, int* nsource, double* source, double* charge, int* nt, double* targ, double* pottarg, double* gradtarg, int* ier); }
+extern int fab(int n);
+extern int isfab(int m);
+
+using namespace LAMMPS_NS;
+using namespace std;
+
+HSMA2D::HSMA2D(LAMMPS* lmp) : KSpace(lmp)
+{
+	maxatom = atom->natoms;
+	me = comm->me;
+	RankID = comm->nprocs;
+	Lx = domain->xprd;
+	Ly = domain->yprd;
+	Lz = domain->zprd * slab_volfactor;
+	PI = 3.141592653589793;
+}
+
+/* ---------------------------------------------------------------------- */
+void HSMA2D::settings(int narg, char** arg)
+{
+	if (narg != 10) error->all(FLERR, "Illegal kspace_style HSMA2D command");
+	tolerance = fabs(utils::numeric(FLERR, arg[0], false, lmp));
+
+	Lambda = utils::numeric(FLERR, arg[1], false, lmp);
+	Gamma= utils::numeric(FLERR, arg[2], false, lmp);
+	p = utils::numeric(FLERR, arg[3], false, lmp);
+	Nw = utils::numeric(FLERR, arg[4], false, lmp);
+	w = utils::numeric(FLERR, arg[5], false, lmp);
+	Fp = utils::numeric(FLERR, arg[6], false, lmp);
+	F = utils::numeric(FLERR, arg[7], false, lmp);
+	IF_FMM_RightTerm = utils::numeric(FLERR, arg[8], false, lmp);
+	IF_FMM_FinalPotential = utils::numeric(FLERR, arg[9], false, lmp);
+
+	if (Lambda < 0)
+		error->all(FLERR, "Lambda shoule be >0.");
+	if (Lambda > 20 || Lambda < 0.2) {
+		error->warning(FLERR, fmt::format("The Lambda is too big or too small! Please use an approximate range of Lambda. Set Lambda to the default value.",
+			update->ntimestep));
+		Lambda = 1.5;
+	}
+	if (p < 1)
+		error->all(FLERR, "p shoule be >=1.");
+	if (p > 50) {
+		error->warning(FLERR, fmt::format("The p is too big! Please use a smaller p. Set p to the default value.",
+			update->ntimestep));
+		p = 6;
+	}
+	if ((IF_FMM_RightTerm) != 0 && (IF_FMM_RightTerm != 1))
+		error->all(FLERR, "Using wrong value of IF_FMM_RightTerm.");
+	if ((IF_FMM_FinalPotential) != 0 && (IF_FMM_FinalPotential != 1))
+		error->all(FLERR, "Using wrong value of IF_FMM_FinalPotential.");
+	if (Fp > F || Fp < 0 || F < 0 || isfab(int(Fp)) != 1 || isfab(int(F)) != 1) {
+		error->warning(FLERR, fmt::format("Using wrong value of Fp and F. Set Fp and F to the default value.",
```

Comment:
Here is another case of using a format call without using the optional argument. There are more. Please either drop the format() and the argument(s) or use the argument.

Reply:
Revised